zrml/prediction-markets/src/benchmarks.rs

```diff
@@ -0,0 +1,402 @@
+        } else {
+            let acc = account("AssetHolder", i, 0);
+            let _ = T::Shares::deposit(fake_asset, &acc, min_liquidity)?;
+        }
+    }
+
+    Ok(())
+}
+
+// Setup a categorical market for fn `internal_resolve`
+fn setup_resolve_common_categorical<T: Config>(
+    acc_total: u32,
+    acc_asset: u32,
+    categories: u16,
+) -> Result<(T::AccountId, T::MarketId), &'static str> {
+    let (caller, marketid) = create_close_and_report_market::<T>(
+        MarketCreation::Permissionless,
+        MarketType::Categorical(categories),
+        Outcome::Categorical(categories.saturating_sub(1)),
+    )?;
+    let _ = generate_accounts_with_assets::<T>(
+        acc_total,
+        acc_asset,
+        Asset::CategoricalOutcome(marketid, categories.saturating_sub(1)),
+    )?;
+    Ok((caller, marketid))
+}
+
+// Setup a categorical market for fn `internal_resolve`
+fn setup_redeem_shares_common<T: Config>(
+    market_type: MarketType,
+) -> Result<(T::AccountId, T::MarketId), &'static str> {
+    let (caller, marketid) =
+        create_market_common::<T>(MarketCreation::Permissionless, market_type.clone())?;
+    let outcome: Outcome;
+
+    if let MarketType::Categorical(categories) = market_type {
+        outcome = Outcome::Categorical(categories.saturating_sub(1));
+    } else if let MarketType::Scalar(range) = market_type {
+        outcome = Outcome::Scalar(range.1);
+    } else {
+        panic!(
+            "setup_redeem_shares_common: Unsupported market type: {:?}",
+            market_type
+        );
+    }
+
+    let _ =
+        Pallet::<T>::do_buy_complete_set(caller.clone(), marketid, MIN_LIQUIDITY.saturated_into())?;
+    let approval_origin = T::ApprovalOrigin::successful_origin();
+    let _ = Call::<T>::admin_move_market_to_closed(marketid)
+        .dispatch_bypass_filter(approval_origin.clone())?;
+    let _ = Call::<T>::report(marketid, outcome)
+        .dispatch_bypass_filter(RawOrigin::Signed(caller.clone()).into())?;
+    let _ = Call::<T>::admin_move_market_to_resolved(marketid)
+        .dispatch_bypass_filter(approval_origin)?;
+    Ok((caller, marketid))
+}
+
+// Setup a scalar market for fn `internal_resolve`
+fn setup_resolve_common_scalar<T: Config>(
+    acc_total: u32,
+    acc_asset: u32,
+) -> Result<(T::AccountId, T::MarketId), &'static str> {
+    let (caller, marketid) = create_close_and_report_market::<T>(
+        MarketCreation::Permissionless,
+        MarketType::Scalar((0u128, u128::MAX)),
+        Outcome::Scalar(u128::MAX),
+    )?;
+    let _ = generate_accounts_with_assets::<T>(
+        acc_total,
+        acc_asset,
+        Asset::ScalarOutcome(marketid, ScalarPosition::Long),
+    )?;
+    Ok((caller, marketid))
+}
+
+benchmarks! {
+    admin_destroy_disputed_market{
+        // a = total accounts
+        let a in 0..10;
+        // b = num. accounts with assets
+        // Unfortunately frame-benchmarking does not allow to b = b.min(a) here
+        let b in 0..10;
```

Comment:
Why only benchmark up to 10 accounts? This seems like a rather low number compared to what could be worst case. I saw the issue you made about the bounds of the account and assets holders, shouldn't we benchmark closer to the bound?

Reply:
I played around a bit with different values (account numbers and repetitions of benchmarks) and found out the following: A higher number increases the benchmark runtime significantly, while increasing the error due to a lower number of repetitions (the data points that are used to approximate the weight function weight(a) are less precise).
I assume that the reason behind this is, that the required weight per account is a linear function of degree 1, i.e. `fn(a) = ba^1 + ca^0`. If this assumption is correct (and the code mirrors that, because its behavior is independent from the actual number of accounts), the first few data points of the curve are already approximating the function and its constant gradient fairly well, provided that these few data points are very precise (many repetitions). If the function is linear and the gradient is well estimated, any value can be derived, up to infinity, assuming that at no point the function fn(a) is non-linear. At least practically (through tests of many different values), this behavior is shown.

Reply:
Makes sense, indeed it should be linear since for each iteration it performs a constant number of storage writes. Can you add this explanation to the documentation inside the benchmark file in case someone reading the code would have this question again in the future?

Reply:
@lsaether done.